tests/reports/unit/test_sonatadata.cpp

```diff
@@ -0,0 +1,127 @@
+            }
+
+            THEN("We check the element ids of the sonata report") {
+                const std::vector<uint32_t> element_ids = sonata->get_element_ids();
+                /*std::vector<uint32_t> compare = { 1000, 1001, 2000, 2001, 2002, 42000, 42001,
+                                                  42002, 42003, 42004, 42005 };*/
+                std::vector<uint32_t> compare = {0, 1, 10, 11, 12, 20, 21, 22, 23, 24, 25};
+                REQUIRE(element_ids == compare);
+            }
+
+            THEN("We check the index pointers of the sonata report") {
+                const std::vector<uint64_t> index_pointers = sonata->get_index_pointers();
+                std::vector<uint64_t> compare = {0, 2, 5};
+                REQUIRE(index_pointers == compare);
+            }
+        }
+        WHEN("We record some other data and prepare the dataset for a small max buffer size") {
+            std::shared_ptr<nodes_t> nodes = std::make_shared<nodes_t>(
+                std::initializer_list<nodes_t::value_type>{{1, node}, {2, node2}, {42, node42}});
+            int num_steps = 3;
+            size_t max_buffer_size = 256;
+            std::unique_ptr<SonataData> sonata2 = std::make_unique<SonataData>(
+                "test_sonatadata2", max_buffer_size, num_steps, dt, tstart, tend, nodes);
+
+            sonata2->prepare_dataset();
+            for (int i = 0; i < num_steps; i++) {
+                sonata2->record_data(i);
+                sonata2->update_timestep(i);
+            }
+            // sonata2->write_data();
+            sonata2->close();
+
+            THEN(
+                "The buffer size is the number of steps to write that fit on the buffer times the "
+                "total elements") {
+                // 256 / sizeof(double) / 11 = 2
+                // buffer_size = 11 * 2
+                REQUIRE(sonata2->get_report_buffer().size() == 22);
+            }
+        }
+    }
+    GIVEN("Spike data") {
+        std::vector<double> spike_timestamps{0.3, 0.1, 0.2, 1.3, 0.7};
+        std::vector<int> spike_node_ids{3, 5, 2, 3, 2};
+        std::unique_ptr<SonataData> sonata_spikes =
+            std::make_unique<SonataData>("spikes", spike_timestamps, spike_node_ids);
+        WHEN("We write the spikes") {
+            sonata_spikes->write_spikes_header();
+            THEN("We check that the spike nodes ids are ordered according to timestamps") {
+                const std::vector<int> node_ids = sonata_spikes->get_spike_node_ids();
+                std::vector<int> compare = {5, 2, 3, 2, 3};
+                REQUIRE(node_ids == compare);
+            }
+            THEN("We check that the spike timestamps are in order") {
+                const std::vector<double> timestamps = sonata_spikes->get_spike_timestamps();
+                std::vector<double> compare = {0.1, 0.2, 0.3, 0.7, 1.3};
+                REQUIRE(timestamps == compare);
+            }
+        }
+        sonata_spikes->close();
```

Comment:
This is checked with the reference h5 files in the integration tests, but I can add also unit tests comparing the file itself (although I was waiting for the readers to compare the input of the writers and the return of the readers).

Reply:
How was the reference file generated?  In other portions of this library, we generate the reference file with completely separate code (`tests/data/generate.py`), so that it's easier to generate new data, and we're exercising different code paths.

Being able to roundtrip data with a reader is also important.

Reply:
👍 

@jorblancoa : simple h5py script to generate reference file would be useful!